Create SBOMs for images pulled during package creation

mikhailswift opened the pull request:
## Description

Creates SBOMs for all images pulled during `zarf package create`. Stores them in a sboms directory within the package.

<!-- Please include a summary of the change. Any relevant motivation or context is also helpful, as well as any dependencies that are required for this change -->

## Related Issue

<!--- This project prefers to accept pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->

Fixes #22

## Type of change

<!-- Please delete options that are not relevant -->
- [x] New feature (non-breaking change which adds functionality)
## Checklist before merging

<!-- Please delete options that are not relevant -->

- [ ] Tests have been added/updated as necessary (add the `needs-tests` label)
- [ ] Documentation has been updated as necessary (add the `needs-docs` label)
- [x] An ADR has been written as necessary (add the `needs-adr` label) [ [1](https://github.com/joelparkerhenderson/architecture-decision-record) [2](https://cognitect.com/blog/2011/11/15/documenting-architecture-decisions) [3](https://adr.github.io/) ]
- [x] (Optional) Changes have been linted locally with [golangci-lint](https://github.com/golangci/golangci-lint). (NOTE: We haven't turned on lint checks in the pipeline yet so linting may be hard if it shows a lot of lint errors in places that weren't touched by changes. Thus, linting is optional right now.)

mikhailswift commented:
Relevant pr: https://github.com/testifysec/witness/pull/167

jeff-mccoy commented on go.mod:
```diff
@@ -5,7 +5,6 @@ go 1.16
 require (
```
We should open a new issue for go update but I'm really just eagerly waiting for .18 and my beautiful generics 😍

mikhailswift replied:
Newly imported code has modules that require go >= 1.17

mikhailswift replied:
Namely:

```
# sigs.k8s.io/json/internal/golang/encoding/json
/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20211208200746-9f7c6b3444d2/internal/golang/encoding/json/encode.go:1249:12: sf.IsExported undefined (type reflect.StructField has no field or method IsExported)
/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20211208200746-9f7c6b3444d2/internal/golang/encoding/json/encode.go:1255:18: sf.IsExported undefined (type reflect.StructField has no field or method IsExported)
note: module requires Go 1.17
```

mikhailswift replied:
And upgrading the zarf module to 1.17 causes unhappiness with opentelemetry 

```
go: finding module for package go.opentelemetry.io/otel/metric/registry
go: finding module for package go.opentelemetry.io/otel/semconv
github.com/defenseunicorns/zarf/test/e2e imports
	github.com/rancher/k3d/v5/cmd/cluster imports
	github.com/rancher/k3d/v5/pkg/client imports
	github.com/rancher/k3d/v5/pkg/runtimes/docker imports
	github.com/docker/cli/cli/command imports
	github.com/docker/cli/cli/trust imports
	github.com/theupdateframework/notary/trustpinning tested by
	github.com/theupdateframework/notary/trustpinning.test imports
	github.com/cloudflare/cfssl/helpers imports
	github.com/google/certificate-transparency-go imports
	go.etcd.io/etcd/v3 imports
	go.etcd.io/etcd/tests/v3/integration imports
	go.etcd.io/etcd/server/v3/embed imports
	go.opentelemetry.io/otel/semconv: module go.opentelemetry.io/otel@latest found (v1.4.1), but does not contain package go.opentelemetry.io/otel/semconv
github.com/defenseunicorns/zarf/test/e2e imports
	github.com/rancher/k3d/v5/cmd/cluster imports
	github.com/rancher/k3d/v5/pkg/client imports
	github.com/rancher/k3d/v5/pkg/runtimes/docker imports
	github.com/docker/cli/cli/command imports
	github.com/docker/cli/cli/trust imports
	github.com/theupdateframework/notary/trustpinning tested by
	github.com/theupdateframework/notary/trustpinning.test imports
	github.com/cloudflare/cfssl/helpers imports
	github.com/google/certificate-transparency-go imports
	go.etcd.io/etcd/v3 imports
	go.etcd.io/etcd/tests/v3/integration imports
	go.etcd.io/etcd/server/v3/embed imports
	go.opentelemetry.io/otel/exporters/otlp imports
	go.opentelemetry.io/otel/sdk/metric/controller/basic imports
	go.opentelemetry.io/otel/metric/registry: module go.opentelemetry.io/otel/metric@latest found (v0.27.0), but does not contain package go.opentelemetry.io/otel/metric/registry
```

Which is addressed by the opentelemetry team here:

https://github.com/open-telemetry/opentelemetry-go/issues/2422#issuecomment-981716733

But... it seems like that's going to require changes somewhere in that dependency change, probably etcd

jeff-mccoy replied:
Okay cool, let's just bump to 1.17 everywhere then properly, no issues going to 1.17 for me.

mikhailswift replied:
Alright, I think I've managed to escape go mod hell and get the weird dependency issues sorted now. Should be bumped to 1.17 now

YrrepNoj commented on cli/internal/packager/create.go:
```diff
@@ -52,7 +53,9 @@ func Create() {
-		images.PullAll(seedImages, tempPath.seedImages)
+		pulledImages := images.PullAll(seedImages, tempPath.seedImages)
+		_ = utils.CreateDirectory(tempPath.sboms, 0700)
+		sbom.CatalogImages(pulledImages, tempPath.sboms)
```
@jeff-mccoy This is writing the sboms to the temp directory which I believe gets blown away when the package creation is finished (unless there is a fatal error). Is there a persistent but intuitive place you think these should be written to instead?

mikhailswift replied:
From my testing the sboms directory is getting packaged into the .zst

YrrepNoj replied:
Oops you're totally correct, I forgot we archive the temp directory into the package itself.

jeff-mccoy commented:
This is going to need some more work before we can merge it.  We will need to have an ADR on the libs we pulled in (it adds multiple deps and increases the binary size by > 10%) as well as a test to verify it's being created.  Also not sure what's going on with the `go.mod`, but it's enormous compared to master.  Finally, we'll need some doc entry to explain what users should see.  ADR won't be needed after the first SBOM PR and we can help with that if needed.

jeff-mccoy commented on cli/internal/sbom/catalog.go:
```diff
@@ -0,0 +1,49 @@
+func CatalogImages(tagToImage map[name.Tag]v1.Image, sbomDir string) {
+	imageCount := len(tagToImage)
+	spinner := message.NewProgressSpinner("Creating SBOMs for %d images.", imageCount)
+	actx, err := attestation.NewContext([]attestation.Attestor{})
```
as a pattern, we always run `defer spinner.Stop()` immediately after invoking `message.NewProgressSpinner()` to ensure there are no spinner leaks.

mikhailswift replied:
Can make that change -- I copied the pattern I saw in images.PullAll

jeff-mccoy replied:
Yeah we have some areas we may have missed that pattern, thx.

jeff-mccoy commented on go.mod:
```diff
@@ -1,37 +1,266 @@
 	sigs.k8s.io/yaml v1.3.0
 )
+
+require (
```
Need to understand what is happening with all these explicit indirect deps.

mikhailswift replied:
Go 1.17 changed the go.mod file to include indirect dependencies: https://go.dev/doc/go1.17 under heading "Pruned module graphs in go 1.17 modules"

jeff-mccoy replied:
TIL 😂 thx.

jeff-mccoy commented on go.mod:
```diff
@@ -1,37 +1,266 @@
-	k8s.io/api v0.22.5
-	k8s.io/apimachinery v0.22.5
-	k8s.io/client-go v0.22.5
+	k8s.io/api v0.23.4
```
whenever we jump k8s versions we need to make sure we're careful to not introducing any version issues with remote clusters, was this bump needed for SBOM or just arbitrary?

mikhailswift replied:
Needed for SBOM -- or at least currently.  Witness imports syft which imports k8s 0.23.0.  I fought with dependencies for a while trying to limit what changed.

jeff-mccoy commented on go.mod:
```diff
@@ -1,37 +1,266 @@
 	github.com/spf13/cobra v1.3.0
 	github.com/stretchr/testify v1.7.0
-	golang.org/x/crypto v0.0.0-20211215165025-cf75a172585e
+	github.com/testifysec/witness v0.1.7-0.20220307231058-f11e00e711dd
```
Would be good to understand the witness versioning plan and how that will be leveraged by zarf in your mind

mikhailswift replied:
Witness will follow SemVer.  We're a bit loose with breaking changes right now due to being in 0.x, which SemVer allow, but we are still making efforts to limit breaking changes.

jeff-mccoy commented on go.mod:
```diff
@@ -1,37 +1,266 @@
 	sigs.k8s.io/kind v0.11.1
-	sigs.k8s.io/kustomize/api v0.8.11
-	sigs.k8s.io/kustomize/kyaml v0.11.0
+	sigs.k8s.io/kustomize/api v0.10.1
```
no issue with kustomize bumps if everything still is happy, but curious was this needed for sbom or not?

mikhailswift replied:
Same as the answer about the k8s api

jeff-mccoy commented on go.mod:
```diff
@@ -1,37 +1,266 @@
+	sigs.k8s.io/structured-merge-diff/v4 v4.2.1 // indirect
+)
+
+replace github.com/derailed/popeye => github.com/testifysec/popeye v0.9.9-0.20220307190025-d441c7496b07
```
Popeye is a useful tool and is used by k9s that we vendor in, running on a fork might become a real problem down the road, can you clarify the need for this?

mikhailswift replied:
A field in a struct in the k8s API changed names, breaking compilation.  It's my goal to get the change upstreamed to popeye and the fork be temporary.

jeff-mccoy reviewed (changes requested):
As already stated, will need ADR/Docs/Tests and items discussed below.  I also noticed SBOM generation can take quite a while, any comments on performance or possible future improvements for that?

mikhailswift commented:
> As already stated, will need ADR/Docs/Tests and items discussed below. I also noticed SBOM generation can take quite a while, any comments on performance or possible future improvements for that?

I've been working on the adr/docs/test, will get something in hsortly.

I can do some tracing to see what part of the process is taking time and identify areas for improvement.

jeff-mccoy commented:
That would be super helpful, thanks!

mikhailswift commented:
Ran some quick tests today profiling sbom generation on `package-example-single-big-bang-package`

Here's a flame graph of the sbom code on a run with a cold image cache

![image](https://user-images.githubusercontent.com/3218582/158244947-9d3e7dad-9504-4dc8-9798-078b3bb5a7ed.png)

The total run time of sbom.CatalogImages was 11.19 seconds.

And here's a flame graph for the same package but with a hot cache

![image](https://user-images.githubusercontent.com/3218582/158246043-f34f66ef-13c9-45d5-bef5-8ea857e4f46d.png)

The total run time of sbom.CatalogImages was 4.57 seconds.

From the flame graphs it appears that it's rebuilding the image tar from the layer cache instead of just using the cache directly as I hoped it would.  This explains why sbom generation is much faster on subsequent runs for the same image since the tar is already built.

Simplest fix will be to try to get the cataloger to use the layer cache directories directly instead of re-building the image's tar file.

mikhailswift commented:
I spent the last bit chasing down performance improvements.  I'll get the ADR and documentation in asap this week while I continue working tests and performance issues.

jeff-mccoy commented:
thanks @mikhailswift note there are some merge conflicts with the tests updates we've merged in recently.

YrrepNoj reviewed (dismissed):
Code changes looks good to me. 

There was a single comment @jeff-mccoy left on adding a quick defer statement for the spinner that's left.

jeff-mccoy commented on docs/adr/0004-generate-sboms-with-witness.md:
```diff
@@ -0,0 +1,19 @@
+
+## Status
+
+Proposed
```
Let's make this accepted now, thx

mikhailswift commented:
Sounds good!

I've been looking into the usage of the sbom a bit, will probably have some questions on the UX of looking at these from Zarf in the short term.

jeff-mccoy commented:
Overall this PR seems reasonable after messing with it for a couple of days, but I have concerns with adding SBOM and slowing down the package create stage without some visible utility to the user.  So I propose we finish and merge this ASAP but not cut a release until we add something to view/use the SBOM.  Here's something I've been playing around with the past couple of days that consumes our SBOM JSON data: https://www.loom.com/share/d85b02b13df442d7a0bae012307db3ee.  Thoughts @mikhailswift?

jeff-mccoy commented on docs/adr/0004-generate-sboms-with-witness.md:
```diff
@@ -0,0 +1,19 @@
+
+## Decision
+
+Using Witness' Syft attestor functionality allows Zarf to continue to get more refined SBOM capabilities as Witness' capabilities expand over time. Syft is capable of finding installed packages and some binaries for statically compiled dependencies over each image within a Zarf package. This allows for SBOMs for each image to be generated and packaged along with the Zarf package.  Capabilities to ...
```
"apabilities to display SBOMs for a Zarf package are forthcoming.", would love for that to have more info.

jeff-mccoy commented on docs/adr/0004-generate-sboms-with-witness.md:
```diff
@@ -0,0 +1,19 @@
+
+## Consequences
+
+Syft has some dependencies that cause some issues currently.  Namely bumping to k8s-api v0.23 causes some incompatibilities with derailed/popeye which has resulted in a fork being made.  This bump may also be undesirable due to compatibility with clusters that do not use this version of the k8s api.  TestifySec is currently investigating removing the Syft dependency entirely and generating the SB ...
```
This needs to be updated now I believe.

jeff-mccoy commented:
This still needs docs/tests, any updates on your end @mikhailswift?

jeff-mccoy commented:
BB Core SBOM, it runs--it's also > 200 MBs 😱🤣

<img width="969" alt="Screen Shot 2022-04-12 at 7 28 21 AM" src="https://user-images.githubusercontent.com/882485/162962512-3aec2db1-7db4-4588-92f9-3403cf22ee99.png">

mikhailswift commented:
> BB Core SBOM, it runs--it's also > 200 MBs screamrofl
> 
> <img alt="Screen Shot 2022-04-12 at 7 28 21 AM" width="969" src="https://user-images.githubusercontent.com/882485/162962512-3aec2db1-7db4-4588-92f9-3403cf22ee99.png">

wew.  something i was considering at the grocery store today is that we could potentially generate the sboms on demand instead of packaging them if the size becomes an issue.  main problem there is that currently syft doesn't support multi-image tar files, and of course it may not be an amazing user experience for the consumer of the sbom.

jeff-mccoy commented:
TIL some ppl daydream about SBOM while getting groceries...

mikhailswift commented:
> TIL some ppl daydream about SBOM while getting groceries...

Send help please

mikhailswift commented:
Okay, I'm doing some last rounds of testing but I believe I've performance of generating the SBOMs improved significantly:

Both of the below timings were measured with .zarf-image-cache erased for a cold run of the single-big-bang-package exampoe

Current code:

```make package-example-single-big-bang-package  25.37s user 4.31s system 38% cpu 1:16.58 total```

Not a precise measurement but watching the timer on the SBOM step seemed to take ~30 seconds 

Changed:

```make package-example-single-big-bang-package  23.31s user 3.43s system 57% cpu 46.313 total```

Same imprecise measurement as above but the SBOM step seemed to take ~10 seconds.

Image pull on both examples took about 30 seconds.

Looking at cpuprofile comparisons again,

Current code:

![image](https://user-images.githubusercontent.com/3218582/163462354-97886175-003f-44f5-959e-bea0513fde78.png)

In this case `image.Read` is taking up ~30% of the cpu time in the CatalogImages function call, and there are some sneaky http and tls calls in there that lead me to believe it could actually be pulling layers over again.

Changed code:
![image](https://user-images.githubusercontent.com/3218582/163462621-8631219f-b7dd-4d7c-b494-b49973160cd2.png)

In this case image.Read is taking ~22% of the cpu time in CatalogImages, but no tls calls happen.

Will have this + docs & tests pushed today

mikhailswift commented:
added some basic docs and updates to the adr.

worked a bit on presentation: https://github.com/defenseunicorns/zarf/commit/1d5d02f0e4533ed5e0df1c9a5f63940234cdecb2

That is very much a hacky work in progress, though

jeff-mccoy commented:
@Madeline-UX
Latest update with the SBOM Viewer thangz

<img width="1012" alt="Screen Shot 202
2-04-26 at 1 51 51 PM" src="https://user-images.githubusercontent.com/882485/165372323-ecf981f8-c004-4631-9fe9-9363361db026.png">
 
<img width="1728" alt="Screen Shot 2022-04-26 at 1 54 00 PM" src="https://user-images.githubusercontent.com/882485/165372330-cf82c955-aa38-4c02-98ff-fc9c637f4d64.png">

YrrepNoj commented on src/internal/sbom/viewer/template.gohtml:
```diff
@@ -0,0 +1,264 @@
+        <div class="element">
+            <svg width="200" height="200" viewBox="0 0 200 200" fill="none" xmlns="http://www.w3.org/2000/svg">
+                <path fill-rule="evenodd" clip-rule="evenodd"
+                    d="M39.3649 39.8969L39.0264 39.678C29.8765 39.489 18.7254 42.4037 6.05101 51.8142C10.0335 55.3755 34.6157 75.1914 64.6042 59.0263C63.8105 58.0754 63.1996 56.9861 62.8022 55.8132C62.4835 54.4006 64.1463 52.3116 65.7393 50.6901C58.6284 43.951 49.299 40.0386 39.5043 39.688L39.3649 39.8969ZM53.9211 180.379C84.1711 189.645 101.366 161.681 101.58 161.332L101.582 161.329C101.707 161. ...
```
out of curiosity, how was this file built/generated?

jeff-mccoy replied:
what do you mean? It's just plain html.  You mean the embedded SVG?

YrrepNoj commented on src/internal/packager/deploy.go:
```diff
@@ -62,9 +63,31 @@ func Deploy() {
+		// Create the directory again
+		utils.CreateDirectory(sbomDir, 0755)
+		for _, file := range sbomViewFiles {
+			// Our file copy lib explodes on these files for some reason...
```
This works so I guess that's fine but I'm curious as to why the copy wouldn't work.. Probably something that's worth  looking into when we have more time.

jeff-mccoy replied:
Yeah it's a bug in that library. We actually only introduced it for the recursive copy feature we really don't use anymore so I'd like to look at moving everything to this pattern (standard lib) and drop that dep soon.
